lib/graphql/subscriptions/anycable_subscriptions.rb

```diff
@@ -123,18 +158,19 @@ def read_subscription(subscription_id)
         end
       end
 
-      # The channel was closed, forget about it.
       def delete_subscription(subscription_id)
-        # Remove subscription ids from all events
-        events = redis.smembers(SUBSCRIPTION_EVENTS_PREFIX + subscription_id)
-        events.each do |event_topic|
-          redis.srem(EVENT_PREFIX + event_topic, subscription_id)
+        events = redis.hget(SUBSCRIPTION_PREFIX + subscription_id, :events)
+        events = events ? JSON.parse(events) : []
+        events.each do |event|
+          redis.srem(FINGERPRINTS_PREFIX + event["topic"], event["fingerprint"])
+          redis.srem(SUBSCRIPTIONS_PREFIX + event["fingerprint"], subscription_id)
```

Comment:
if `SUBSCRIPTIONS_PREFIX + event["fingerprint"]` ends up being empty it might be worth it to remove the fingerprint from `FINGERPRINTS_PREFIX + event["topic"]`. What do you think?

Reply:
Oh, I see. We shouldn't always remove fingerprint from topic fingerprints, but only if there is no more subscriptions for this fingerprint. And there will be issues with concurrent subscriptions and unsubscriptions 😖

Reply:
Replaced set of fingerprints by topic with sorted set, which stores number of subscriptions for every fingerprint value in its score. And removes elements when their score reaches zero (no more subscriptions)